pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,375 @@
+	// chosen to just avoid finding the IP.
+	return gostatsd.UnknownIP, nil
+}
+
+// NewProviderFromViper returns a new k8s provider.
+func NewProviderFromViper(v *viper.Viper, logger logrus.FieldLogger, version string) (gostatsd.CloudProvider, error) {
+	k := util.GetSubViper(v, "k8s")
+	setViperDefaults(k, version)
+
+	// Set up the k8s client
+	clientset, err := createKubernetesClient(k.GetString(ParamUserAgent), k.GetString(ParamKubeconfigPath),
+		k.GetString(ParamKubeconfigContext), k.GetFloat64(ParamAPIQPS), k.GetFloat64(ParamAPIQPSBurst))
+	if err != nil {
+		return nil, err
+	}
+
+	return NewProvider(
+		logger,
+		clientset,
+		PodInformerOptions{
+			ResyncPeriod: k.GetDuration(ParamResyncPeriod),
+			WatchCluster: k.GetBool(ParamWatchCluster),
+			NodeName:     k.GetString(ParamNodeName),
+		},
+		regexp.MustCompile(k.GetString(ParamAnnotationTagRegex)),
+		regexp.MustCompile(k.GetString(ParamLabelTagRegex)))
+}
+
+// NewProvider returns a new k8s provider.
+func NewProvider(logger logrus.FieldLogger, clientset kubernetes.Interface, podInfOpts PodInformerOptions,
+	annotationRegex, labelRegex *regexp.Regexp) (gostatsd.CloudProvider, error) {
+
+	// This list operation is for debugging purposes and failing fast. If this fails then the cache will likely fail.
+	_, err := clientset.CoreV1().Pods(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// If we're not watching the entire cluster we need to limit our watch to pods with our node name
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+	if !podInfOpts.WatchCluster {
+		if podInfOpts.NodeName == "" {
+			return nil, fmt.Errorf("watch-cluster set to false, and node name not supplied")
+		}
+		fieldSelector := "spec.nodeName=" + podInfOpts.NodeName
+		logger.WithField("fieldSelector", fieldSelector).Debug("set fieldSelector for informers")
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fieldSelector
+		}
+	}
+
+	// Create a shared informer factory that watches the correct nodes
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		clientset,
+		podInfOpts.ResyncPeriod,
+		informers.WithTweakListOptions(customWatchOptions))
+
+	// Set up the pod informer which fills an index with the pods we care about
+	podsInf := factory.Core().V1().Pods().Informer()
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+	err = podsInf.AddIndexers(indexers)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
+	// TODO: we should emit events to the k8s API to fit in with the k8s ecosystem
+
+	return &Provider{
+		logger:          logger,
+		podsInf:         podsInf,
+		factory:         factory,
+		annotationRegex: annotationRegex,
+		labelRegex:      labelRegex,
+	}, nil
+}
+
+func (p *Provider) Run(ctx context.Context) {
+	go p.factory.Start(ctx.Done()) // fork
```

Comment:
The factory is annoying because it's impossible to wait until all the goroutines it spawns finish. And the reason for that is that `Start()` does not block, it spawns informers' `Run()` methods in goroutines and returns control. 
I suggest you directly construct the informer and use it here OR if you decide to stick with the factory at least remove that `go` statement as it's not needed.

Reply:
What I've done here is make the Run() synchronous (i.e. we don't call it with a goroutine when spawning a cloud provider - see handler_cloud.go:318). This way we can asynchronously start whatever we want inside this context, and block until the cache returns. This ensures that the lookup can never happen before the k8s provider is ready, and we'll actually get a good debugging trace if the cache won't populate for some reason.

WDYT about the approach? It did feel a little weird changing the delineation point for asynchronous calls.

Reply:
`Run()` is supposed to block until the context is cancelled and when it returns, the "thing" is not supposed to be running and hence is not supposed to be used (this is not documented on the interface, sorry). In this PR the semantics of the contract are being changed and only in a single place. Rather than introducing an exception (and an undocumented one especially), maybe introduce another method to the `CloudProvider` interface that allows waiting for the provider to become ready. Something like `WaitUntilReady(context.Context) error` that blocks until provider is ready to be used. This is explicit and clear.
On the other hand, when gostatsd is redeployed all producers are not going to be waiting as there is no backpressure in UDP. And even for TCP, the socket hasn't been opened at this point yet so...
So regardless of what we do the metrics are either going to be dropped or sent untagged. I guess we have that choice here. We can design the thing to block/delay metrics processing for up to N seconds and then either drop metrics for which the tags couldn't be obtained or send them though down the pipeline. Can be configurable even.
So far the approach used has been to block/delay metrics for a very short while (while cache is populated for the IP) or not block at all and then push them further. I think we should go one step further and control how much blocking/waiting we want. We'll be able to do it after my refactoring (not in the first PR). If we agree on this design we can just leave things without any blocking/waiting and just do the smarts later.

p.s. another thing to consider - what if informer has not yet synced fully but already has the needed information? If done via `WaitForCacheToSync()` things will be blocked for longer than necessary. Via blocking index-like functionality we can have smarter waiting rather than all or nothing.

Reply:
I think we should just leave it and do the smarts later, as you say. Totally agree on the contract stuff, thanks for clarifying.

The main thing that was attracting me to waiting for the cache sync was debugging, since if the cache does not initially sync then something is clearly wrong. I wonder if there's a way to surface that sort of error nicely but asynchronously. I guess we could put a context with timeout in the run method, but just not block on it. Then if it times out we can print an error message saying it took a long time to sync the cache and maybe there is a problem.

Thoughts?

Reply:
I think the initial sync problems vs sync problems that may happen later shouldn't be viewed as separate things. If what we have here works we can keep it for now and I'll refactor this in my PR.

Or add a method to the interface. That is a nice way to surface problems with initial sync.

Reply:
I've just set it back to async calls as per the contract and removed cache checking. We can revisit this later.